### PR TITLE
Add new commands and integrate undo/redo functionality

### LINKED ISSUE #45: Unable to "make" a recipe
as a user, i can "make" a recipe so that the app can track and account for the ingredients used, and record statistics.

### LINKED ISSUE #68: Implement undo/redo
Undo/redo has been implemented for AB3, but needs to be modified to work with our project. Several things to note:

- All commands need to unambigously specify how the model changes (cannot be delegated to other classes), or else the `HistoryManager` will be unable to know what to reverse when undoing.
- Specifically, undoing the completion of recipes is likely to be problematic, since the current idea is to remove the oldest ingredients in the ingredient book when completing a recipe, which doesn't work well with above.

### LINKED ISSUE #88: Cannot edit ingredients

### LINKED ISSUE #92: Allow removing of ingredient by index
Removing ingredients by a given quantity currently only works with names, but it should probably work with indexes as well, to align with deleting ingredients.

### PR DESCRIPTION
**Pull request by seowalex**
Adds new commands for making recipes and deleting specific quantities of ingredients. Also integrated undo/redo functionality into all commands that modify the model.

Closes #45, closes #68, closes #88 and closes #92.

### CODE REVIEW

File: src/main/java/chopchop/logic/commands/MakeRecipeCommand.java [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+            + "Example: " + COMMAND_WORD + " 1";
+
+    public static final String MESSAGE_MAKE_RECIPE_SUCCESS = "Recipe made: %s";
+    public static final String MESSAGE_MAKE_RECIPE_ERROR = "Cannot make recipe due to ingredient '%s'";
+    public static final String MESSAGE_RECIPE_NOT_FOUND = "No recipe named '%s'";
+    public static final String MESSAGE_INGREDIENT_NOT_FOUND = "No ingredient named '%s'";
+
+    private final ItemReference item;
+    // (Original ingredient, (Consumed ingredient, Remaining ingredient))
+    private List<Pair<Ingredient, Pair<Ingredient, Ingredient>>> ingredients;
```
Comment (zhiayang) [inline src/main/java/chopchop/logic/commands/MakeRecipeCommand.java:38]: maybe you should make a smol nested helper class for this

File: src/main/java/chopchop/model/ingredient/Ingredient.java [hunk 2]
```diff
@@ -69,25 +68,24 @@ public Quantity getQuantity() {
...
             return this.sets.firstEntry().getValue();
         }
     }
 
     public Optional<ExpiryDate> getExpiryDate() {
+        assert !this.sets.isEmpty();
+
         // just return the first expiry date.
-        return this.sets.firstEntry().getKey();
+        return this.sets.firstKey();
```
Comment (zhiayang) [inline src/main/java/chopchop/model/ingredient/Ingredient.java:85]: pog

**Review by zhiayang — COMMENTED** [review]
looks good

also why tf you got something against using `var`?

File: src/main/java/chopchop/logic/commands/MakeRecipeCommand.java (hunk 1 above)
Comment (seowalex) [reply src/main/java/chopchop/logic/commands/MakeRecipeCommand.java:39]: Was lazy, will do.

File: src/main/java/chopchop/model/ingredient/Ingredient.java (hunk 2 above)
Comment (seowalex) [reply src/main/java/chopchop/model/ingredient/Ingredient.java:85]: Just OCD don't mind me.

**Comment by seowalex** [comment]
Still missing tests, but functionality should all be there.

**Comment by zhiayang** [comment]
btw doing `closes A, B, C, and D` only marks `A` as being closed by this pr

**Comment by seowalex** [comment]
TIL ok will change.

**Pull request merged**